linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,350 @@
+
+`threadpool_create` shall create a SRW lock by calling `srw_lock_create`.
+
+`threadpool_create` shall create a shared semaphore between threads with initialized value zero.
+
+`insert_idx`  and `consume_idx` for the task array shall be intialzied to 0.
+
+If any error occurs, `threadpool_create` shall fail and return `NULL`.
+
+### threadpool_destroy
+
+```C
+MOCKABLE_FUNCTION(, void, threadpool_destroy, THREADPOOL_HANDLE, threadpool);
+```
+
+`threadpool_destroy` frees the resouces associated with `threadpool`.
+
+If `threadpool` is `NULL`, `threadpool_destroy` shall return.
+
+`threadpool_destroy` shall implicit close if `threadpool` is open. 
+
+`threadpool_destroy` shall free the resources associated with the threadpool handle. 
+
+`threadpool_destroy` shall destroy the semphore by calling `sem_destroy`.
+
+`threadpool_destroy` shall destroy the SRW lock by calling `srw_lock_destroy`.
+
+### threadpool_open_async
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_open_async, THREADPOOL_HANDLE, threadpool, ON_THREADPOOL_OPEN_COMPLETE, on_open_complete, void*, on_open_complete_context);
+```
+
+`threadpool_open_async` opens the threadpool asynchronously.
+
+If `threadpool` is `NULL`, `threadpool_open_async` shall fail and return a non-zero value. 
+
+`threadpool_open_async` shall call `sm_open_begin`. 
+
+`threadpool_open_async` shall create the threads for `threadpool` using `ThreadAPI_Create`.
+
+If `threadpool` has already been opened, `threadpool_open_async` shall fail and return a non-zero value.
+
+If one of the thread creation fails, `threadpool_open_async` shall fail and return a non-zero value, terminate all threads already created.
+
+Otherwise, `threadpool_open_async` shall shall call `sm_open_end` with true for success.
+
+`threadpool_open_async` shall succeed and return zero, indicate open success to the user by calling the `on_open_complete` callback with `THREADPOOL_OPEN_OK`.
+
+### threadpool_close
+
+```C
+MOCKABLE_FUNCTION(, void, threadpool_close, THREADPOOL_HANDLE, threadpool);
+```
+
+`threadpool_close` closes the threadpool.
+
+If `threadpool` is `NULL`, `threadpool_close` shall fail and return.
+
+Otherwise, `threadpool_close` shall join all threads in the `threadpool`.
+
+`threadpool_close` shall call `sm_close_begin`.
+
+`threadpool_close` shall call `sm_close_end`.
+
+### threadpool_schedule_work
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work, THREADPOOL_HANDLE, threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
+```
+
+`threadpool_schedule_work` schedule the threadpool to work.
+
+If `threadpool` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+If `work_function` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall get the threadpool state by calling `interlocked_add`.
```

Comment:
>get the threadpool state by calling `interlocked_add`. [](http://example.com/codeflow?start=33&length=54)

Nope, no more of this. We use sm, so it means we call sm_exec_begin now.

Reply:
So there is no checking of any state. There should be no state maintained by threadpool_linux. It is all maintained by sm now.
If you have 2 states (one maintained by threadpool_linux and one maintained by sm) there is no way to change both of them atomically.

Reply:
removed